core/src/main/java/smile/projection/ICA.java

```diff
@@ -0,0 +1,365 @@
+ * You may obtain a copy of the License at
+ *  
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package smile.projection;
+
+import java.io.Serializable;
+import smile.math.Math;
+import smile.math.matrix.Matrix;
+import smile.stat.distribution.GaussianDistribution;
+import smile.math.matrix.DenseMatrix;
+import smile.math.matrix.EVD;
+
+/**
+ * Independent Component Analysis (ICA) is a computational method for separating
+ * a multivariate signal into additive subcomponents.
+ * <p>
+ * We use FastICA for this implementation, which is an efficient and popular
+ * algorithm for independent component analysis invented by Aapo Hyvärinen at
+ * Helsinki University of Technology.
+ * <p>
+ * Like most ICA algorithms, FastICA seeks an orthogonal rotation of prewhitened
+ * data, through a fixed-point iteration scheme, that maximizes a measure of
+ * non-Gaussianity of the rotated components.
+ * <p>
+ * Non-gaussianity serves as a proxy for statistical independence, which is a
+ * very strong condition and requires infinite data to verify. FastICA can also
+ * be alternatively derived as an approximative Newton iteration.
+ * <p>
+ * Typical use case is 'cocktail party problem' where to identify original
+ * speakers from mixed audio recordings.
+ * 
+ * <h2>References</h2>
+ * <ol>
+ * <li>Aapo Hyvärinen: Fast and robust fixed-point algorithms for independent
+ * component analysis, 1999</li>
+ * <li>Aapo Hyvärinen, Erkki Oja: Independent component analysis: Algorithms and
+ * applications, 2000</li>
+ * </ol>
+ * 
+ * @see PCA
+ * 
+ * @author rayeaster
+ */
+public class ICA implements Projection<double[]>, Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * The dimension of feature space.
+     */
+    private int p;
+    /**
+     * The dimension of input space.
+     */
+    private int n;
+    /**
+     * constant to check if convergence on one component.
+     */
+    private double tolerance;
+    /**
+     * constant for maximum iteration for one component search.
+     */
+    private int maxIter;
+    /**
+     * The projected sample mean.
+     */
+    private double[] pmu;
+    /**
+     * Projection matrix.
+     */
+    private DenseMatrix projection;
+
+    /**
+     * determine the functions to calculate derivative and second-order derivative
+     */
+    private NegEntropyFunc funcMode = NegEntropyFunc.LOGCOSH;
+
+    /**
+     * functions to calculate derivative and second-order derivative
+     */
+    public enum NegEntropyFunc {
```

Comment:
It is better to define an interface that has `f` (nonquadratic nonlinear function), `g` (first derivative) and `g2` (second derivative). And provide implementations for `log cosh` and `- exp(-u^2/2)`. This way the user can easily plug in their special functions without changing this code.

Reply:
Make this interface and `log cosh` etc implementations inner classes of ICA.